internal/controller/controllers/clusterdeployments_controller.go

```diff
@@ -1929,31 +1973,57 @@ func clusterSpecSynced(cluster *hiveext.AgentClusterInstall, syncErr error) {
 	})
 }
 
-func clusterRequirementsMet(clusterInstall *hiveext.AgentClusterInstall, status string, registeredHosts, approvedHosts, unsyncedHosts int) {
+func clusterRequirementsMet(
+	clusterInstall *hiveext.AgentClusterInstall,
+	cluster *common.Cluster,
+	status string,
+	approvedHosts,
+	unsyncedHosts int,
+	clusterAPI cluster.API,
+) error {
 	var condStatus corev1.ConditionStatus
 	var reason string
 	var msg string
 
 	switch status {
 	case models.ClusterStatusReady:
-		expectedHosts := clusterInstall.Spec.ProvisionRequirements.ControlPlaneAgents +
-			clusterInstall.Spec.ProvisionRequirements.WorkerAgents
-		if registeredHosts < expectedHosts {
+		expectedMasterCount := clusterInstall.Spec.ProvisionRequirements.ControlPlaneAgents
+		expectedWorkerCount := clusterInstall.Spec.ProvisionRequirements.WorkerAgents
+
+		// We count directly from the DB because the hosts are not loaded, and we want to avoid loading all hosts
+		mastersCountPtr, err := clusterAPI.GetHostCountByRole(*cluster.ID, models.HostRoleMaster, true)
+		if err != nil {
+			return err
+		}
+
+		workersCountPtr, err := clusterAPI.GetHostCountByRole(*cluster.ID, models.HostRoleWorker, true)
+		if err != nil {
+			return err
+		}
```

Comment:
Is it possible to do this logic outside of this function? If it errors here, then we won't see any status updates on the AgentClusterInstall, but I think we should always try to set status updates

Reply:
Makes sense. We need to know the amount of masters / workers in order to show more accurate message, so we need to count during `updateStatus`. We can take it one level up to `updateStatus` (as I see we already have some functions returning errors there), but I am not sure it will help because in case of error we will still not show any status no ? WDYT ?

Reply:
I think taking it up one level up is a good idea. If there's an error, could we log it but still also update status?

Reply:
you think it would be better to update status even though the other functions one level up return 
```
return ctrl.Result{Requeue: true}, nil
```
?

I can maybe show this error as the status message in case of error, WDYT ?
I mean in the case of error -

```
condStatus = corev1.ConditionFalse
reason = hiveext.ClusterInsufficientAgentsReason
msg = "failed to count the amount of control plane agents and worker agents bound to the cluster"
```

Reply:
Yeah showing an error message would be better! I just don't think it should fail setting the status if the function GetHostCountByRole errors out

Reply:
Ack.

Reply:
I implemented the changes, Let me know what you think.